**BÀI TẬP CHƯƠNG 2**

**23174600124_TẠ HỒNG HOÀ_DHKL17A3HN**

Kết nối đến SQLite và tạo bảng student và bảng course

In [48]:
import sqlite3
import pandas as pd

In [49]:
conn = sqlite3.connect("students.db") 
cursor = conn.cursor()

In [50]:
cursor.execute("DROP TABLE IF EXISTS student;")
cursor.execute("DROP TABLE IF EXISTS course;")

cursor.execute("""
CREATE TABLE student (
    student_id INTEGER PRIMARY KEY,
    name TEXT,
    class TEXT,
    course_id INTEGER,
    score REAL
);
""")

cursor.execute("""
CREATE TABLE course (
    id INTEGER PRIMARY KEY,
    course_name TEXT
);
""")

conn.commit()

In [51]:
student_data = [
    (1, "Nguyen Minh Hoang", "May Tinh", 12, 6.7),
    (2, "Tran Thi Lan", "Kinh Te", 34, 9.2),
    (3, "Pham Van Nam", "Toan Tin", None, 7.9),   
    (4, "Le Thanh Huyen", "Toan Tin", 20, 7.2),
    (5, "Vu Quoc Anh", "May Tinh", 24, 8.0),
    (6, "Dang Thuy Linh", "May Tinh", 24, 5.5),
    (7, "Bui Tien Dung", "Kinh Te", 34, 9.2),
    (8, "Ho Ngoc Mai", "Toan Tin", 20, 8.8),
    (9, "Duong Huu Phuc", "Kinh Te", None, 7.2),   
    (10, "Cao Thi Hanh", "May Tinh", None, 7.0)     
]

course_data = [
    (12, "Giai tich"),
    (34, "Thong ke"),
    (26, "Tin hoc")
]

cursor.executemany("INSERT INTO student VALUES (?, ?, ?, ?, ?);", student_data)

cursor.executemany("INSERT INTO course VALUES (?, ?);", course_data)

conn.commit()


In [52]:
student_df = pd.read_sql_query("SELECT * FROM student;", conn)
print("Bảng Student:")
display(student_df)

course_df = pd.read_sql_query("SELECT * FROM course;", conn)
print("Bảng Course:")
display(course_df)

Bảng Student:


,student_id,name,class,course_id,score
0,1,Nguyen Minh Hoang,May Tinh,12.0,6.7
1,2,Tran Thi Lan,Kinh Te,34.0,9.2
2,3,Pham Van Nam,Toan Tin,NaN,7.9
3,4,Le Thanh Huyen,Toan Tin,20.0,7.2
4,5,Vu Quoc Anh,May Tinh,24.0,8.0
5,6,Dang Thuy Linh,May Tinh,24.0,5.5
6,7,Bui Tien Dung,Kinh Te,34.0,9.2
7,8,Ho Ngoc Mai,Toan Tin,20.0,8.8
8,9,Duong Huu Phuc,Kinh Te,NaN,7.2
9,10,Cao Thi Hanh,May Tinh,NaN,7.0


Bảng Course:


,id,course_name
0,12,Giai tich
1,26,Tin hoc
2,34,Thong ke


Nhận xét:

Đối với bảng Student:

- Bảng gồm 10 dòng và 5 cột.
- Các cột bao gồm:
    + student_id: mã sinh viên (kiểu số nguyên),
    + name: tên sinh viên (kiểu chuỗi),
    + class: lớp (kiểu chuỗi),
    + course_id: mã môn học (kiểu số thực, có tồn tại giá trị thiếu),
    + score: điểm số (kiểu số thực).
- Cột course_id có 3 giá trị bị thiếu (NaN) tương ứng với các sinh viên có student_id là 3, 9 và 10. Điều này có thể gây ảnh hưởng đến quá trình xác định môn học mà các sinh viên này.

Đối với bảng Course:

- Bảng gồm 3 dòng và 2 cột.
- Các cột bao gồm:
    + id: mã môn học (kiểu số nguyên),
    + course_name: tên môn học (kiểu chuỗi).

**1. Hãy kết nối hai bảng trên theo những cách sau:**
**- Sử dụng tích Decartes.**
**- Sử dụng JOIN: INNER JOIN, LEFT JOIN, RIGHT JOIN, FULL OUTER JOIN.**

Sử dụng tích Decartes

In [53]:
print("Tích Decartes:")
cursor.execute('''
SELECT * FROM student, course;
''')
df_decartes = pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score', 'id', 'course_name'])
print(df_decartes)
print("\n")


Tích Decartes:
    student_id               name     class  course_id  score  id course_name
0            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12   Giai tich
1            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  26     Tin hoc
2            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  34    Thong ke
3            2       Tran Thi Lan   Kinh Te       34.0    9.2  12   Giai tich
4            2       Tran Thi Lan   Kinh Te       34.0    9.2  26     Tin hoc
5            2       Tran Thi Lan   Kinh Te       34.0    9.2  34    Thong ke
6            3       Pham Van Nam  Toan Tin        NaN    7.9  12   Giai tich
7            3       Pham Van Nam  Toan Tin        NaN    7.9  26     Tin hoc
8            3       Pham Van Nam  Toan Tin        NaN    7.9  34    Thong ke
9            4     Le Thanh Huyen  Toan Tin       20.0    7.2  12   Giai tich
10           4     Le Thanh Huyen  Toan Tin       20.0    7.2  26     Tin hoc
11           4     Le Thanh Huyen  Toan Tin      

Kết luận:

- Bảng kết quả thu được là từ phép tích Descartes giữa hai bảng students và courses, do đó mỗi sinh viên được kết hợp với tất cả các môn học và xuất hiện lặp lại nhiều lần.
- Với 10 sinh viên và 3 môn học, tổng số dòng của bảng kết quả là 10 × 3 = 30 dòng.
- Việc xuất hiện các giá trị NaN ở cột course_id cho thấy vẫn có những sinh viên chưa đăng ký môn học, tuy nhiên họ vẫn được liệt kê trong kết quả của tích Descartes.
- Giá trị của course_id và score không hoàn toàn tương ứng với id trong bảng courses, điều này cho thấy không phải tất cả sinh viên đều tham gia đầy đủ các môn học.
- Cột score bị lặp lại theo từng môn học đối với mỗi sinh viên, phản ánh rằng dữ liệu chưa được tổng hợp hoặc liên kết một cách chính xác theo từng môn học cụ thể.

Sử dụng JOIN: INNER JOIN, LEFT JOIN, RIGHT JOIN, FULL OUTER JOIN

INNER JOIN

In [54]:
print("INNER JOIN:")
cursor.execute('''
SELECT * FROM student INNER JOIN course ON student.course_id = course.id;
''')
df_inner = pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score', 'id', 'course_name'])
print(df_inner)
print("\n")

INNER JOIN:
   student_id               name     class  course_id  score  id course_name
0           1  Nguyen Minh Hoang  May Tinh         12    6.7  12   Giai tich
1           2       Tran Thi Lan   Kinh Te         34    9.2  34    Thong ke
2           7      Bui Tien Dung   Kinh Te         34    9.2  34    Thong ke




Kết luận:

- Bảng chỉ hiển thị các dòng có giá trị khớp giữa course_id của bảng students và id của bảng courses.
- Chỉ có 3 sinh viên xuất hiện, cho thấy số lượng sinh viên có đăng ký môn học trùng khớp giữa hai bảng là khá ít.

LEFT JOIN

In [55]:
print("LEFT JOIN:")
cursor.execute('''
SELECT * FROM student LEFT JOIN course ON student.course_id = course.id;
''')
df_left = pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score', 'id', 'course_name'])
print(df_left)
print("\n")

LEFT JOIN:
   student_id               name     class  course_id  score    id course_name
0           1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12.0   Giai tich
1           2       Tran Thi Lan   Kinh Te       34.0    9.2  34.0    Thong ke
2           3       Pham Van Nam  Toan Tin        NaN    7.9   NaN         NaN
3           4     Le Thanh Huyen  Toan Tin       20.0    7.2   NaN         NaN
4           5        Vu Quoc Anh  May Tinh       24.0    8.0   NaN         NaN
5           6     Dang Thuy Linh  May Tinh       24.0    5.5   NaN         NaN
6           7      Bui Tien Dung   Kinh Te       34.0    9.2  34.0    Thong ke
7           8        Ho Ngoc Mai  Toan Tin       20.0    8.8   NaN         NaN
8           9     Duong Huu Phuc   Kinh Te        NaN    7.2   NaN         NaN
9          10       Cao Thi Hanh  May Tinh        NaN    7.0   NaN         NaN




Kết luận:

- LEFT JOIN giữ lại toàn bộ các bản ghi từ bảng students và chỉ bổ sung dữ liệu từ bảng courses khi có giá trị tương ứng.
- Các giá trị NaN ở cột id và course_name cho thấy có nhiều sinh viên có course_id không khớp với id trong bảng courses.

RIGHT JOIN

In [56]:
print("RIGHT JOIN:")
cursor.execute('''
SELECT * FROM course LEFT JOIN student ON course.id = student.course_id;
''')
df_right = pd.DataFrame(cursor.fetchall(), columns=['id', 'course_name', 'student_id', 'name', 'class', 'course_id', 'score'])
print(df_right)
print("\n")

RIGHT JOIN:
   id course_name  student_id               name     class  course_id  score
0  12   Giai tich         1.0  Nguyen Minh Hoang  May Tinh       12.0    6.7
1  26     Tin hoc         NaN                NaN       NaN        NaN    NaN
2  34    Thong ke         7.0      Bui Tien Dung   Kinh Te       34.0    9.2
3  34    Thong ke         2.0       Tran Thi Lan   Kinh Te       34.0    9.2




Kết luận:
- RIGHT JOIN giữ lại toàn bộ các bản ghi từ bảng courses và chỉ bổ sung thông tin từ bảng students khi có giá trị khớp.
- Xuất hiện dòng có course_name là "Tin hoc" nhưng các cột thông tin sinh viên đều là NaN, điều này cho thấy không có sinh viên nào đăng ký môn học này hoặc dữ liệu đăng ký còn thiếu.

FULL OUTER JOIN (mô phỏng bằng UNION của LEFT JOIN và RIGHT JOIN)

In [57]:
query = '''
SELECT * FROM student LEFT JOIN course 
ON student.course_id = course.id
UNION ALL
SELECT * FROM course LEFT JOIN student 
ON course.id = student.course_id
WHERE student.course_id IS NULL;
'''
df = pd.read_sql_query(query, conn)
print("FULL OUTER JOIN: ", df)


FULL OUTER JOIN:      student_id               name     class  course_id  score    id  \
0            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12.0   
1            2       Tran Thi Lan   Kinh Te       34.0    9.2  34.0   
2            3       Pham Van Nam  Toan Tin        NaN    7.9   NaN   
3            4     Le Thanh Huyen  Toan Tin       20.0    7.2   NaN   
4            5        Vu Quoc Anh  May Tinh       24.0    8.0   NaN   
5            6     Dang Thuy Linh  May Tinh       24.0    5.5   NaN   
6            7      Bui Tien Dung   Kinh Te       34.0    9.2  34.0   
7            8        Ho Ngoc Mai  Toan Tin       20.0    8.8   NaN   
8            9     Duong Huu Phuc   Kinh Te        NaN    7.2   NaN   
9           10       Cao Thi Hanh  May Tinh        NaN    7.0   NaN   
10          26            Tin hoc       NaN        NaN    NaN   NaN   

   course_name  
0    Giai tich  
1     Thong ke  
2          NaN  
3          NaN  
4          NaN  
5          NaN  
6     Thon

Kết luận:
- FULL OUTER JOIN giữ lại toàn bộ các bản ghi từ cả hai bảng students và courses, đồng thời điền giá trị NaN tại những vị trí không có sự khớp nối.
- Xuất hiện dòng có id = 26 và course_name = "Tin hoc" nhưng không có thông tin sinh viên tương ứng, điều này cho thấy không có sinh viên nào đăng ký môn học này hoặc dữ liệu đăng ký chưa đầy đủ.

**2. Hãy cập nhật những giá trị course_id còn thiếu trong bảng student bằng câu lệnh SQL, trong đó các giá trị được điền là những giá trị nằm trong bảng course và loại bỏ những bản ghi tham gia những môn học không tồn tại bảng course. Sau đó hãy cho biết:**

**a. Tổng số sinh viên, điểm trung bình của từng lớp.**

**b. Tổng số sinh viên, điểm trung bình của từng môn học.**

**c. Phân loại thi đua theo số điểm của từng môn học biết:** 

**-   i. Điểm TB ≥  9.0: Xuất sắc.** 
**-   ii. 6.0 ≤ Điểm TB ≤ 8.9: Tốt.** 
**-  iii. Điểm TB < 6.0: Kém.** 

In [58]:
cursor.execute('''
UPDATE student
SET course_id = (
    SELECT id FROM course ORDER BY RANDOM() LIMIT 1
)
WHERE course_id IS NULL;
''')
conn.commit()

cursor.execute('''
DELETE FROM student
WHERE course_id NOT IN (SELECT id FROM course);
''')
conn.commit()

cursor.execute('SELECT * FROM student;')
print(pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score']))


   student_id               name     class  course_id  score
0           1  Nguyen Minh Hoang  May Tinh         12    6.7
1           2       Tran Thi Lan   Kinh Te         34    9.2
2           3       Pham Van Nam  Toan Tin         12    7.9
3           7      Bui Tien Dung   Kinh Te         34    9.2
4           9     Duong Huu Phuc   Kinh Te         12    7.2
5          10       Cao Thi Hanh  May Tinh         12    7.0


Kết luận:
- Các dòng có course_id là NaN trước đó đã được thay thế bằng các mã môn học được lấy từ bảng course.
- Những bản ghi có course_id không tồn tại trong bảng course đã được loại bỏ.
- Bảng kết quả sau xử lý chỉ còn lại các sinh viên có course_id hợp lệ và không còn xuất hiện giá trị NaN.

a. Tổng số sinh viên, điểm trung bình của từng lớp.

In [59]:
query = '''
SELECT class, COUNT(student_id) AS total_students, ROUND(AVG(score), 2) AS avg_score
FROM student
GROUP BY class;
'''

df = pd.read_sql_query(query, conn)
print(df)


      class  total_students  avg_score
0   Kinh Te               3       8.53
1  May Tinh               2       6.85
2  Toan Tin               1       7.90


Kết luận:
- Kinh Te: 8.53 — Điểm trung bình cao nhất trong các lớp, cho thấy chất lượng học tập tốt.
- May Tinh: 6.85 — Điểm trung bình thấp nhất, cần cải thiện chất lượng học tập.
- Toan Tin: 7.90 — Ở mức khá, có thể duy trì và cải thiện thêm. Tuy nhiên, lớp Toan Tin chỉ có 1 sinh viên nên kết quả này có thể chưa phản ánh đầy đủ tính đại diện của điểm trung bình.

b. Tổng số sinh viên, điểm trung bình của từng môn học. 

In [60]:
query = '''
SELECT course.course_name, COUNT(student.student_id) AS total_students, ROUND(AVG(student.score), 2) AS avg_score
FROM student
LEFT JOIN course ON student.course_id = course.id
GROUP BY course.course_name;
'''

df = pd.read_sql_query(query, conn)
print(df)


  course_name  total_students  avg_score
0   Giai tich               4        7.2
1    Thong ke               2        9.2


Kết luận:
- Thong ke là môn học có kết quả tốt nhất với điểm trung bình cao.
- Giai tich có thể cần được cải thiện cả về số lượng sinh viên đăng ký lẫn chất lượng học tập.

c. Phân loại thi đua theo số điểm của từng môn học biết: 

- i. Điểm TB ≥  9.0: Xuất sắc. 
- ii. 6.0 ≤ Điểm TB ≤ 8.9: Tốt. 
- iii. Điểm TB < 6.0: Kém.

In [61]:
query = '''
SELECT course.course_name,
       ROUND(AVG(student.score), 2) AS avg_score,
       CASE
           WHEN ROUND(AVG(student.score), 2) >= 9.0 THEN 'Xuất sắc'
           WHEN ROUND(AVG(student.score), 2) BETWEEN 6.0 AND 8.9 THEN 'Tốt'
           ELSE 'Kém'
       END AS rating
FROM student
LEFT JOIN course ON student.course_id = course.id
GROUP BY course.course_name;
'''

df = pd.read_sql_query(query, conn)
print(df)


  course_name  avg_score    rating
0   Giai tich        7.2       Tốt
1    Thong ke        9.2  Xuất sắc


Kết luận:
- Môn Thong ke dẫn đầu với xếp loại Xuất sắc và có điểm trung bình cao nhất.
- Giai tich được xếp loại Tốt, tuy nhiên vẫn cần có các giải pháp để tiếp tục nâng cao chất lượng.
- Không có môn học nào bị xếp loại Kém, đây là một tín hiệu tích cực phản ánh chất lượng đào tạo hiện tại.

**3. Hãy xếp hạng sinh viên thông qua:** 

**a. Điểm số.** 

**b. Điểm số theo lớp học.** 

**c. Điểm số theo mã môn học.** 

**và cho biết top 3 sinh viện đạt thứ hạng cao nhất, top 3 sinh viên đạt thứ hạng thấp nhất theo 
từng trường hợp trên.**

a. Điểm số.

In [62]:
query = '''
SELECT student_id, name, class, course_id, score,
       RANK() OVER (ORDER BY score DESC) AS rank
FROM student
ORDER BY score DESC;
'''

df_rank = pd.read_sql_query(query, conn)
print(df_rank)


   student_id               name     class  course_id  score  rank
0           2       Tran Thi Lan   Kinh Te         34    9.2     1
1           7      Bui Tien Dung   Kinh Te         34    9.2     1
2           3       Pham Van Nam  Toan Tin         12    7.9     3
3           9     Duong Huu Phuc   Kinh Te         12    7.2     4
4          10       Cao Thi Hanh  May Tinh         12    7.0     5
5           1  Nguyen Minh Hoang  May Tinh         12    6.7     6


Kết luận:

- Lớp Kinh Te có thành tích nổi bật với cả hạng 1 và hạng 4, cho thấy chất lượng học tập tương đối ổn định.
- Lớp May Tinh có xu hướng đạt điểm thấp hơn so với các lớp khác, cần xem xét cải thiện phương pháp giảng dạy.
- Không có sinh viên nào có điểm dưới 6.5, điều này cho thấy mặt bằng học lực chung ở mức khá tốt.

b. Điểm số theo lớp học. 

In [63]:
query = '''
SELECT student_id, name, class, course_id, score,
       RANK() OVER (PARTITION BY class ORDER BY score DESC) AS class_rank
FROM student
ORDER BY class, class_rank;
'''

df_class_rank = pd.read_sql_query(query, conn)
print(df_class_rank)


   student_id               name     class  course_id  score  class_rank
0           2       Tran Thi Lan   Kinh Te         34    9.2           1
1           7      Bui Tien Dung   Kinh Te         34    9.2           1
2           9     Duong Huu Phuc   Kinh Te         12    7.2           3
3          10       Cao Thi Hanh  May Tinh         12    7.0           1
4           1  Nguyen Minh Hoang  May Tinh         12    6.7           2
5           3       Pham Van Nam  Toan Tin         12    7.9           1


Kết luận:

- Lớp Kinh Te có mức độ cạnh tranh cao, thể hiện qua khoảng cách điểm số tương đối lớn giữa các thứ hạng.
- Lớp May Tinh cần nâng cao chất lượng học tập do điểm số của sinh viên đứng đầu chưa thực sự nổi bật.
- Lớp Toan Tin có mặt bằng điểm ở mức khá, tuy nhiên không có cạnh tranh, cần thu hút thêm sinh viên để tăng tính cạnh tranh.

c. Điểm số theo mã môn học. 

In [64]:
query = '''
SELECT student_id, name, class, course_id, score,
       RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS course_rank
FROM student
ORDER BY course_id, course_rank;
'''

df_course_rank = pd.read_sql_query(query, conn)
print(df_course_rank)


   student_id               name     class  course_id  score  course_rank
0           3       Pham Van Nam  Toan Tin         12    7.9            1
1           9     Duong Huu Phuc   Kinh Te         12    7.2            2
2          10       Cao Thi Hanh  May Tinh         12    7.0            3
3           1  Nguyen Minh Hoang  May Tinh         12    6.7            4
4           2       Tran Thi Lan   Kinh Te         34    9.2            1
5           7      Bui Tien Dung   Kinh Te         34    9.2            1


Kết luận:

Với course_id = 12 (Giai tich):
- Phạm Văn Nam đạt 7.9 điểm (hạng 1), là sinh viên có kết quả cao nhất trong môn này.
- Các sinh viên còn lại gồm Dương Hữu Phúc (7.2 điểm), Cao Thị Hạnh (7.0 điểm) và Nguyễn Minh Hoàng (6.7 điểm) có điểm thấp hơn, cho thấy mặt bằng điểm ở mức trung bình – khá và cần được cải thiện thêm.
Với course_id = 34 (Thong ke):
- Trần Thị Lan và Bùi Tiến Dũng cùng đạt 9.2 điểm và đồng hạng 1.
- Đây là môn có kết quả nổi bật nhất với điểm số cao, thể hiện chất lượng học tập tốt.

Cho biết top 3 sinh viện đạt thứ hạng cao nhất, top 3 sinh viên đạt thứ hạng thấp nhất theo từng trường hợp trên.

Top 3 cao nhất & thấp nhất theo điểm số

In [65]:
query = '''
WITH ranked_students AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER (ORDER BY score DESC) AS rank_desc,
           RANK() OVER (ORDER BY score ASC) AS rank_asc
    FROM student
)
SELECT * FROM ranked_students
WHERE rank_desc <= 3 OR rank_asc <= 3
ORDER BY rank_desc;
'''
df_top_all = pd.read_sql_query(query, conn)
print(" Top 3 cao nhất & thấp nhất theo điểm số toàn bộ:")
print(df_top_all)
print("\n")


 Top 3 cao nhất & thấp nhất theo điểm số toàn bộ:
   student_id               name     class  course_id  score  rank_desc  \
0           2       Tran Thi Lan   Kinh Te         34    9.2          1   
1           7      Bui Tien Dung   Kinh Te         34    9.2          1   
2           3       Pham Van Nam  Toan Tin         12    7.9          3   
3           9     Duong Huu Phuc   Kinh Te         12    7.2          4   
4          10       Cao Thi Hanh  May Tinh         12    7.0          5   
5           1  Nguyen Minh Hoang  May Tinh         12    6.7          6   

   rank_asc  
0         5  
1         5  
2         4  
3         3  
4         2  
5         1  




Top 3 cao nhất:
- Trần Thị Lan và Bùi Tiến Dũng (9.2 điểm) — Hạng 1.
- Phạm Văn Nam (7.9 điểm) — Hạng 3.

Top 3 thấp nhất:
- Nguyễn Minh Hoàng (6.7 điểm) — Hạng 6 (thấp nhất).
- Cao Thị Hạnh (7.0 điểm) — Hạng 5.
- Dương Hữu Phúc (7.2 điểm) — Hạng 4.

Top 3 cao nhất & thấp nhất theo điểm số từng lớp

In [66]:
query = '''
WITH ranked_students AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER (PARTITION BY class ORDER BY score DESC) AS class_rank_desc,
           RANK() OVER (PARTITION BY class ORDER BY score ASC) AS class_rank_asc
    FROM student
)
SELECT * FROM ranked_students
WHERE class_rank_desc <= 3 OR class_rank_asc <= 3
ORDER BY class, class_rank_desc;
'''

df_top_class = pd.read_sql_query(query, conn)
print(" Top 3 cao nhất & thấp nhất theo điểm số từng lớp:")
print(df_top_class)
print("\n")


 Top 3 cao nhất & thấp nhất theo điểm số từng lớp:
   student_id               name     class  course_id  score  class_rank_desc  \
0           2       Tran Thi Lan   Kinh Te         34    9.2                1   
1           7      Bui Tien Dung   Kinh Te         34    9.2                1   
2           9     Duong Huu Phuc   Kinh Te         12    7.2                3   
3          10       Cao Thi Hanh  May Tinh         12    7.0                1   
4           1  Nguyen Minh Hoang  May Tinh         12    6.7                2   
5           3       Pham Van Nam  Toan Tin         12    7.9                1   

   class_rank_asc  
0               2  
1               2  
2               1  
3               2  
4               1  
5               1  




 Top 3 cao nhất & thấp nhất theo điểm số từng lớp
 
Lớp Kinh Tế:
- Cao nhất: Trần Thị Lan và Bùi Tiến Dũng (9.2 điểm) — Đồng hạng 1.
- Thấp nhất: Dương Hữu Phúc (7.2 điểm) — Hạng 3.

Lớp Máy Tính:
- Cao nhất: Cao Thị Hạnh (7.0 điểm) — Hạng 1.
- Thấp nhất: Nguyễn Minh Hoàng (6.7 điểm) — Hạng 2 (thấp nhất).

Lớp Toán Tin:
- Cao nhất: Phạm Văn Nam (7.9 điểm) — Hạng 1.
- Thấp nhất: Cũng là Phạm Văn Nam (7.9 điểm) — Hạng 1 (lớp chỉ có 1 sinh viên).

Top 3 cao nhất & thấp nhất theo điểm số từng mã môn học

In [67]:
query = '''
WITH ranked_students AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS course_rank_desc,
           RANK() OVER (PARTITION BY course_id ORDER BY score ASC) AS course_rank_asc
    FROM student
)
SELECT * FROM ranked_students
WHERE course_rank_desc <= 3 OR course_rank_asc <= 3
ORDER BY course_id, course_rank_desc;
'''

df_top_course = pd.read_sql_query(query, conn)
print(" Top 3 cao nhất & thấp nhất theo điểm số từng mã môn học:")
print(df_top_course)
print("\n")


 Top 3 cao nhất & thấp nhất theo điểm số từng mã môn học:
   student_id               name     class  course_id  score  \
0           3       Pham Van Nam  Toan Tin         12    7.9   
1           9     Duong Huu Phuc   Kinh Te         12    7.2   
2          10       Cao Thi Hanh  May Tinh         12    7.0   
3           1  Nguyen Minh Hoang  May Tinh         12    6.7   
4           2       Tran Thi Lan   Kinh Te         34    9.2   
5           7      Bui Tien Dung   Kinh Te         34    9.2   

   course_rank_desc  course_rank_asc  
0                 1                4  
1                 2                3  
2                 3                2  
3                 4                1  
4                 1                1  
5                 1                1  




Kết luận:

Với course_id = 12 (Giai tich):
- Top cao nhất: Phạm Văn Nam (7.9 điểm) — hạng 1; tiếp theo là Dương Hữu Phúc (7.2 điểm) và Cao Thị Hạnh (7.0 điểm).
- Top thấp nhất: Nguyễn Minh Hoàng (6.7 điểm) — hạng thấp nhất; tiếp theo là Cao Thị Hạnh (7.0 điểm) và Dương Hữu Phúc (7.2 điểm).
- Nhìn chung, điểm số của môn này ở mức trung bình – khá, chưa có sinh viên đạt mức xuất sắc.
Với course_id = 34 (Thong ke):
- Top cao nhất: Trần Thị Lan và Bùi Tiến Dũng cùng đạt (9.2 điểm) — đồng hạng 1.
- Top thấp nhất: Cũng chính hai sinh viên này (9.2 điểm) do số lượng ít và không có sự chênh lệch điểm số.
- Điều này cho thấy môn học có kết quả rất tốt nhưng thiếu sự phân hóa về điểm số giữa các sinh viên.

**4. Hãy bổ sung thêm một trường graduation_date có kiểu dữ liệu là DATETIME vào bảng student để xác định thời gian tốt nghiệp của từng bạn, trong đó thời gian tốt nghiệp được xác định bởi thời gian hiện tại cộng với số hạng tương ứng của bạn đó tính theo điểm số.**

Thêm cột graduation_date vào bảng student

In [68]:
cursor.execute('''
ALTER TABLE student ADD COLUMN graduation_date DATETIME;
''')
conn.commit()

Cập nhật graduation_date dựa trên thứ hạng

In [69]:
cursor.execute('''
WITH ranked_students AS (
    SELECT student_id, RANK() OVER (ORDER BY score DESC) AS rank
    FROM student
)
UPDATE student
SET graduation_date = DATETIME('now', '+' || (
    SELECT rank FROM ranked_students WHERE ranked_students.student_id = student.student_id
) || ' months');
''')
conn.commit()


 Kiểm tra lại bảng student với trường graduation_date

In [70]:
query = '''
SELECT student_id, name, class, course_id, score, graduation_date
FROM student
ORDER BY score DESC;
'''

df_updated = pd.read_sql_query(query, conn)
print(" Bảng `student` sau khi thêm trường `graduation_date`:")
print(df_updated)


 Bảng `student` sau khi thêm trường `graduation_date`:
   student_id               name     class  course_id  score  \
0           2       Tran Thi Lan   Kinh Te         34    9.2   
1           7      Bui Tien Dung   Kinh Te         34    9.2   
2           3       Pham Van Nam  Toan Tin         12    7.9   
3           9     Duong Huu Phuc   Kinh Te         12    7.2   
4          10       Cao Thi Hanh  May Tinh         12    7.0   
5           1  Nguyen Minh Hoang  May Tinh         12    6.7   

       graduation_date  
0  2026-05-03 13:40:11  
1  2026-05-03 13:40:11  
2  2026-07-03 13:40:11  
3  2026-08-03 13:40:11  
4  2026-09-03 13:40:11  
5  2026-10-03 13:40:11  


Bảng student sau khi thêm trường graduation_date đã xác định được thời gian tốt nghiệp của từng sinh viên. Cụ thể:

Trần Thị Lan và Bùi Tiến Dũng (Kinh Tế):
- Điểm: 9.2 — Ngày tốt nghiệp: 08/04/2025.

Phạm Văn Nam (Toán Tin):
- Điểm: 7.9 — Ngày tốt nghiệp: 08/06/2025.

Dương Hữu Phúc (Kinh Tế):
- Điểm: 7.2 — Ngày tốt nghiệp: 08/07/2025.

Cao Thị Hạnh (Máy Tính):
- Điểm: 7.0 — Ngày tốt nghiệp: 08/08/2025.

Nguyễn Minh Hoàng (Máy Tính):
- Điểm: 6.7 — Ngày tốt nghiệp: 08/09/2025.


Thời gian tốt nghiệp giữa các sinh viên có sự chênh lệch rõ ràng, phản ánh sự khác biệt về kết quả học tập.
Việc bổ sung trường này giúp dễ dàng theo dõi và phân loại sinh viên theo tiến độ tốt nghiệp dựa trên năng lực học tập.